### ISSUE TITLE
MVC sln in non-responsive state while rebuilding and solution closure 

### ISSUE THREAD

**Issue by jinujoseph** [NuGet/Home#4637]
**Repro steps**

- Clone mvc repro github.com/aspnet/Mvc/commit/d3cce192925ff41d0d79aba3728ab1ff80962a49
- Open mvc.sln 
Either rebuild or close the solution and it will put VS in a non-responsive state.
Captured both dmp 

\\mlangfs1\public\Jinuj\dmp\mvc_rebuild_hang.DMP 
\\mlangfs1\public\Jinuj\dmp\mvc_unloading.DMP 

@ManishJayaswal  has looked into this dmps 

First case looks like , Nuget restore is on the UI thread
```
mscorlib.dll!System.Threading.Monitor.Wait(object obj, int millisecondsTimeout, bool exitContext)	Unknown
 	mscorlib.dll!System.Threading.Monitor.Wait(object obj, int millisecondsTimeout)	Unknown
 	mscorlib.dll!System.Threading.ManualResetEventSlim.Wait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.SpinThenBlockingWait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.InternalWait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.Wait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.Wait(System.TimeSpan timeout)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskFactory.WaitSynchronouslyCore(System.Threading.Tasks.Task task)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskFactory.WaitSynchronously(System.Threading.Tasks.Task task)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTask.CompleteOnCurrentThread()	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTask<bool>.CompleteOnCurrentThread()	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskFactory.Run<bool>(System.Func<System.Threading.Tasks.Task<bool>> asyncMethod, Microsoft.VisualStudio.Threading.JoinableTaskCreationOptions creationOptions)	Unknown
 	NuGet.SolutionRestoreManager.dll!NuGet.SolutionRestoreManager.SolutionRestoreWorker.Restore(NuGet.PackageManagement.VisualStudio.SolutionRestoreRequest request)	Unknown
 	NuGet.SolutionRestoreManager.dll!NuGet.SolutionRestoreManager.RestoreManagerPackage.BuildEvents_OnBuildBegin(EnvDTE.vsBuildScope scope, EnvDTE.vsBuildAction Action)	Unknown 
```

saw a spike in memory too 

![Uploading a.JPG…]()

Issue 2 looks like (large number of background threads are blocked on NuGet SolutionRestoreManager)

 ```
	mscorlib.dll!System.Threading.Monitor.Wait(object obj, int millisecondsTimeout, bool exitContext) Line 203	C#
 	mscorlib.dll!System.Threading.Monitor.Wait(object obj, int millisecondsTimeout) Line 213	C#
 	mscorlib.dll!System.Threading.ManualResetEventSlim.Wait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 645	C#
 	mscorlib.dll!System.Threading.Tasks.Task.SpinThenBlockingWait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 3327	C#
 	mscorlib.dll!System.Threading.Tasks.Task.InternalWait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 3259	C#
 	mscorlib.dll!System.Threading.Tasks.Task.Wait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 3167	C#
 	mscorlib.dll!System.Threading.Tasks.Task.Wait(System.TimeSpan timeout) Line 3083	C#
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskFactory.WaitSynchronouslyCore(System.Threading.Tasks.Task task)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskFactory.WaitSynchronously(System.Threading.Tasks.Task task)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTask.CompleteOnCurrentThread()	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTask<bool>.CompleteOnCurrentThread()	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskFactory.Run<bool>(System.Func<System.Threading.Tasks.Task<bool>> asyncMethod, Microsoft.VisualStudio.Threading.JoinableTaskCreationOptions creationOptions)	Unknown
 	NuGet.PackageManagement.VisualStudio.dll!NuGet.PackageManagement.VisualStudio.VSSolutionManager.IsSolutionFullyLoaded.get()	Unknown
 	NuGet.SolutionRestoreManager.dll!NuGet.SolutionRestoreManager.SolutionRestoreWorker.ScheduleRestoreAsync(NuGet.PackageManagement.VisualStudio.SolutionRestoreRequest request, System.Threading.CancellationToken token)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<bool>.Start<NuGet.SolutionRestoreManager.SolutionRestoreWorker.<ScheduleRestoreAsync>d__31>(ref NuGet.SolutionRestoreManager.SolutionRestoreWorker.<ScheduleRestoreAsync>d__31 stateMachine) Line 471	C#
 	NuGet.SolutionRestoreManager.dll!NuGet.SolutionRestoreManager.SolutionRestoreWorker.ScheduleRestoreAsync(NuGet.PackageManagement.VisualStudio.SolutionRestoreRequest request, System.Threading.CancellationToken token)	Unknown
 	NuGet.SolutionRestoreManager.dll!NuGet.SolutionRestoreManager.VsSolutionRestoreService.NominateProjectAsync(string projectUniqueName, NuGet.SolutionRestoreManager.IVsProjectRestoreInfo projectRestoreInfo, System.Threading.CancellationToken token)	Unknown
>	Microsoft.VisualStudio.ProjectSystem.Managed.VS.dll!Microsoft.VisualStudio.ProjectSystem.VS.NuGet.NuGetRestorer.ProjectPropertyChangedAsync.AnonymousMethod__0() Line 140	C#
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskFactory.RunAsync(System.Func<System.Threading.Tasks.Task> asyncMethod, bool synchronouslyBlocking, Microsoft.VisualStudio.Threading.JoinableTaskCreationOptions creationOptions, System.Delegate entrypointOverride)	Unknown
 	Microsoft.VisualStudio.ProjectSystem.Managed.VS.dll!Microsoft.VisualStudio.ProjectSystem.VS.NuGet.NuGetRestorer.ProjectPropertyChangedAsync(System.Tuple<System.Collections.Immutable.ImmutableList<Microsoft.VisualStudio.ProjectSystem.IProjectValueVersions>, System.Collections.Immutable.IImmutableDictionary<Microsoft.VisualStudio.ProjectSystem.NamedIdentity, System.IComparable>> sources) Line 137	C#
```

 

build:26215.1.d15rel

**Comment by emgarten** [comment]
@jainaashish @alpaix is this stack trace covered by one of the other hang bugs already?

**Comment by jainaashish** [comment]
It looks more like #4420 and should be fixed with d15rel 26216.0 n beyond releases.

**Comment by mishra14** [comment]
@jainaashish : Could you validate this please. If fixed, then close.

**Comment by jainaashish** [comment]
@nkolev92 was working on this last week... He can share his findings. If I'm not wrong he was able to repro it with or without our fix, So it isn't easy to repro anyways

**Comment by nkolev92** [comment]
I was **not** able to repro this with and without the fix. 
So as @jainaashish said, it's a hard repro.

**Comment by jinujoseph** [comment]
i have a VM with the repo, if you want to take a look at.

**Comment by nkolev92** [comment]
With @jainaashish's help I managed to repro & verify with d15rel.26216.0 
The root cause was solved in #4420
